File: swati_app/custom_script/item/item.py
import frappe


@frappe.whitelist()
def get_next_item_code(custom_type, custom_category, custom_subcategory):
	prefix = "{0}-{1}-{2}-".format(
		custom_type.strip(), custom_category.strip(), custom_subcategory.strip()
	)
	existing = frappe.get_all(
		"Item",
		filters={"item_code": ["like", prefix + "%"]},
		pluck="item_code"
	)
	max_count = 0
	for code in existing:
		suffix = code[len(prefix):]
		if suffix.isdigit():
			max_count = max(max_count, int(suffix))
	return "{0}{1:04d}".format(prefix, max_count + 1)


def validate(doc, method):
    duplicate = frappe.db.exists(
        "Item",
        {
            "item_name": doc.item_name,
            "name": ["!=", doc.name]
        }
    )

    if duplicate:
        frappe.throw(
            f"Item Name <b>{doc.item_name}</b> already exists ({duplicate}). Please use a unique Item Name."
        )

File: swati_app/custom_script/item/item.js
frappe.ui.form.on('Item', {
	custom_type: function(frm) {
		generate_item_code(frm);
	},
	custom_category: function(frm) {
		generate_item_code(frm);
	},
	custom_subcategory: function(frm) {
		generate_item_code(frm);
	}
});

function generate_item_code(frm) {
	if (!frm.is_new()) return;

	if (frm.doc.custom_type && frm.doc.custom_category && frm.doc.custom_subcategory) {
		frappe.call({
			method: 'swati_app.custom_script.item.item.get_next_item_code',
			args: {
				custom_type: frm.doc.custom_type,
				custom_category: frm.doc.custom_category,
				custom_subcategory: frm.doc.custom_subcategory
			},
			callback: function(r) {
				if (r.message) {
					frm.set_value('item_code', r.message);
				}
			}
		});
	} else {
		// Clear if any part is removed
		frm.set_value('item_code', '');
	}
}